### PR TITLE
Table index updated

### PR DESCRIPTION
**Pull request by daniel-sol**
Similar to pull request table_index, just merged with the recent updates of main.
Have removed that a table index is required for all tables. 
There are tables that are pure lists, and do not need an index.
Now the method that checked for table index just removes it if it is None.

### CODE REVIEW

File: src/fmu/dataio/_definitions.py [hunk 1]
```diff
@@ -61,6 +61,12 @@ def __post_init__(self):
     "timeseries": None,
 }
 
+STANDARD_TABLE_COLUMNS = {
```
Comment (jcrivenaes) [inline src/fmu/dataio/_definitions.py:64]: It refers to INDEX columns only, so perhaps better with `STANDARD_TABLES_INDEX_COLUMNS`?

File: src/fmu/dataio/_objectdata_provider.py [hunk 2]
```diff
@@ -469,6 +481,52 @@ def _derive_spec_bbox_arrowtable(self):
...
+        if "REAL" in columns:
+            index.append("REAL")
+
+        return index
+
+    def _check_index(self, index):
+        """Check the table index
+        ARGS:
+            index (list): list of column names
+
```
Comment (jcrivenaes) [inline src/fmu/dataio/_objectdata_provider.py:521]: Punctuation ending first line (which should be more precise, then space, then Args: not ARGS: for google docstring:

And please, use typing annotations

def _check_index(self, index: list):
```
"""Check if valid tablindex is present.

     Args:
            index: list of column names
```

File: src/fmu/dataio/_objectdata_provider.py [hunk 3]
```diff
@@ -469,6 +481,52 @@ def _derive_spec_bbox_arrowtable(self):
...
+        Raises:
+            KeyError: if index contains names that are not in self
+        """
+        # Using generator, this is lazy
+        unwanteds = (item for item in index if item not in self._get_columns())
+        for unwanted in unwanteds:
+            raise KeyError(f"{unwanted} is not in table")
```
Comment (jcrivenaes) [inline src/fmu/dataio/_objectdata_provider.py:528]: The logic that "unwanted not in a table" should raise an error is not obvious?

File: src/fmu/dataio/_objectdata_provider.py [hunk 4]
```diff
@@ -477,9 +535,10 @@ def _derive_timedata(self):
             return {}
 
         if self.dataio.legacy_time_format:
-            return self._derive_timedata_legacy()
+            timedata = self._derive_timedata_legacy()
         else:
-            return self._derive_timedata_newformat()
+            timedata = self._derive_timedata_newformat()
+        return timedata
```
Comment (jcrivenaes) [inline src/fmu/dataio/_objectdata_provider.py:541]: Why extra code here?

File: src/fmu/dataio/dataio.py [hunk 5]
```diff
@@ -531,6 +530,7 @@ class ExportData:
     verbosity: str = "CRITICAL"
     vertical_domain: dict = field(default_factory=dict)
     workflow: str = ""
+    table_index: Union[None, list] = None
```
Comment (jcrivenaes) [inline src/fmu/dataio/dataio.py:533]: `Union[None, list] = None` can be shorten with `Optional[list] = None`

File: src/fmu/dataio/dataio.py [hunk 6]
```diff
@@ -751,6 +751,15 @@ def _check_obj_if_file(self, obj: Any) -> Any:
...
         return obj
 
+    def _check_table_index(self):
+        """Removing table_index if None"""
+        try:
+            if self._metadata["data"]["table_index"] is None:
+                del self._metadata["data"]["table_index"]
+                logger.debug("Table index was none, removed")
+        except KeyError:
+            logger.debug("No table index in the first place")
```
Comment (jcrivenaes) [inline src/fmu/dataio/dataio.py:761]:
```
if not self._metadata["data"].get("table_index"):
    del self._metadata["data"]["table_index"]
    logger.debug("Table index was none, removed")
else:
     logger.debug("No table index in the first place")
```
If the class is table, should it be more than `logger.debug()` ? E.g. a `UserWarning`?

File: src/fmu/dataio/dataio.py [hunk 7]
```diff
@@ -828,6 +838,8 @@ def export(self, obj, return_symlink=False, **kwargs) -> str:
         Returns:
             String: full path to exported item.
         """
+        self.table_index = kwargs.get("table_index", self.table_index)
+        print(self.table_index)
```
Comment (jcrivenaes) [inline src/fmu/dataio/dataio.py:842]: To be removed?

File: tests/conftest.py [hunk 8]
```diff
@@ -24,10 +24,13 @@
 else:
     HAS_PYARROW = True
 
+import fmu.dataio as dio
+from fmu.config import utilities as ut
 from fmu.dataio.dataio import ExportData, read_metadata
 
 logger = logging.getLogger(__name__)
 
+logger.setLevel("ERROR")
```
Comment (jcrivenaes) [inline tests/conftest.py:33]: I think we should avoid setting logging level. See also https://github.com/equinor/fmu-dataio/issues/270

**Review by jcrivenaes — CHANGES_REQUESTED** [review]
I think the table index feature can be quite useful 👍 . Consider comments, and then we can a second around afterwards.

File: src/fmu/dataio/dataio.py (hunk 6 above)
Comment (perolavsvendsen) [reply src/fmu/dataio/dataio.py:761]: > ```
> if not self._metadata["data"].get("table_index"):
>     del self._metadata["data"]["table_index"]
>     logger.debug("Table index was none, removed")
> else:
>      logger.debug("No table index in the first place")
> ```

Possibly redundant comment if everything is going away anyway (common removal of Nones later), but first line above will evaluate to True both if table_index is None and if it's not there. But the del operation will fail with KeyError if it is not there.

Suggest something like:
```
if "table_index" in self._metadata["data"] and self._metadata["data"]["table_index"] is None:
    del self._metadata["data"]["table_index"]
```

...if it is kept.

Comment (jcrivenaes) [reply src/fmu/dataio/dataio.py:761]: My bad, just delete the "not" in the if... and it should work

File: src/fmu/dataio/_definitions.py (hunk 1 above)
Comment (daniel-sol) [reply src/fmu/dataio/_definitions.py:64]: Agree, just note the slight modification, changed to STANDARD_TABLE_INDEX_COLUMNS

File: src/fmu/dataio/_objectdata_provider.py (hunk 2 above)
Comment (daniel-sol) [reply src/fmu/dataio/_objectdata_provider.py:521]: About the dockstring: Yes, not sure why that all caps snuck in, gone now. When it comes to the dot, if that is what you want, that is fine

Comment (daniel-sol) [reply src/fmu/dataio/_objectdata_provider.py:521]: When it comes to the unwanteds: changed to not_founds, to make it clearer what they are. The idea is to make the user aware that he/she is trying to use column names that are not present in the table

Comment (daniel-sol) [reply src/fmu/dataio/_objectdata_provider.py:521]: Woops: last comment belongs to issue below: look at it there

File: src/fmu/dataio/_objectdata_provider.py (hunk 4 above)
Comment (daniel-sol) [reply src/fmu/dataio/_objectdata_provider.py:541]: Sorry, came up in the linting process, couldn't resist.
_objectdata_provider.py:479:8: R1705: Unnecessary "else" after "return", remove the "else" and de-indent the code inside it (no-else-return) (pylint 2.15.5): But no skin off my nose to remove it if you that is what you want)

File: src/fmu/dataio/_objectdata_provider.py (hunk 3 above)
Comment (daniel-sol) [reply src/fmu/dataio/_objectdata_provider.py]: Dunno where the comment ended up, could have sworn that I had commented on this before. But the idea is that the user should not be allowed to use as table index column names that are not present in table. Have also renamed unwanteds to not_founds to make code more self explanatory.

File: src/fmu/dataio/dataio.py (hunk 6 above)
Comment (daniel-sol) [reply src/fmu/dataio/dataio.py]: Now removed

File: src/fmu/dataio/_objectdata_provider.py (hunk 4 above)
Comment (jcrivenaes) [reply src/fmu/dataio/_objectdata_provider.py:541]: ok, not a big deal

**Pull request merged**